### PR TITLE
Add last commit resource optimization feature for datasources

### PR DESCRIPTION
**Pull request by graben**
@gytis : Could you plz have a look on this. Works as expected if using pooled connections but behave odd for generic connection because of LRCO is skipped by IsSameRMOverrideXAResource wrapping. Tests are green but might cause issues in failure scenarios?

### CODE REVIEW

**Comment by graben** [comment]
Maybe remove recovery ITs as they don't really make sense because 2phase commit exception is the weak moment of lrco. Those tests fails most time under Linux but not under Windows. hmm. :-|

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        private PooledJdbcConnection() {
+            this.closed = false;
+        }
+
+        @Override
+        public void close() throws SQLException {
+            if (!this.closed) {
+                try {
+                    LrcoXAConnection.this.physicalConnection.rollback();
+                    LrcoXAConnection.this.physicalConnection.setAutoCommit(true);
```
Comment (gytis) [inline narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java:124]: Why do you need to set autocomit to true here?

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    @Override
+    public XAResource getXAResource() throws SQLException {
+        return new LrcoXAResource(this.physicalConnection);
+    }
+
+    @Override
+    public Connection getConnection() throws SQLException {
+        Connection lastHandleConnection = this.handleConnection;
```
Comment (gytis) [inline narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java:64]: Could you explain a bit this method? What's the purpose of the PooledJdbcConnection class? Other than close it seems to always delegate to the physical connection.

**Review by gytis — COMMENTED** [review]
What do you mean by remove recovery ITs? What do the existing recovery tests have to do with LRCO PR?

**Comment by graben** [comment]
> What do you mean by remove recovery ITs? What do the existing recovery tests have to do with LRCO PR?

My first commit included instances of GenericRecoveryIT tests for lrco non-pool and pool version. But they did not work very well and after thinking about I removed them from this PR. So it was something new that disappeared. ;)

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java (hunk 1 above)
Comment (graben) [reply narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java:124]: This is something most lrco impls I read are doing after reset original connection to set autocommit back to true. Some even saves old content.

File: narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java (hunk 2 above)
Comment (graben) [reply narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java:64]: That's the contract from JDBC API:

Creates and returns a <code>Connection</code> object that is a handle
for the physical connection that
this <code>PooledConnection</code> object represents.
The connection pool manager calls this method when an application has
called the method <code>DataSource.getConnection</code> and there are
no <code>PooledConnection</code> objects available. See the
{@link PooledConnection interface description} for more information.

**Comment by gytis** [comment]
Hmm, do you remember what did not work in a recovery scenario? While LRCO resource is excluded from the transaction log, all the previous resources should be logged and recovered in case of a failure.

**Comment by graben** [comment]
Well, maybe it's correlated with the fact that for non pooled resource the LRCO is skipped by IsSameRMOverrideXAResource wrapping. But I need to reover the test case to get the Exception again.

**Comment by graben** [comment]
image

Runs well under Windows with JDK 8, fails under Linux with JDK8, but seems to work with JDK11. Hmm...

**Comment by gytis** [comment]
Could you add it to this PR, so that I could look into it?

**Comment by graben** [comment]
@gytis : I add the first one, and it fails as expected.

**Comment by graben** [comment]
Will have a look after rebase to junit5...

**Comment by graben** [comment]
No better with junit5

**Comment by gytis** [comment]
@graben I've added one commit with a recovery test change for LRCO. We shouldn't expect a database record to exist, because the LRCO resource record is not written and therefore shouldn't be recovered.

**Comment by graben** [comment]
Hope you're right because in the non-pooled way the optimization is not working at all because of a Narayana "bug" with IsSameRMOverrideXAResource.  This way the full 2PC protocol is used with an empty prepare. Only the pooled variant really goes the internal LRCO way in Narayana. A pooled test case this way still fails. Or am I wrong?

**Comment by gytis** [comment]
I'll try that out as well and discuss it with the Narayana team.

**Comment by graben** [comment]
@gytis : I add a small fix to make output of integration tests better readable.

**Comment by graben** [comment]
image

Just for the record. If real LRCO is used commit is done instead of prepare. That's why for pooled LRCO DataSource resource the entry is committed before simulated crash, yet!

**Comment by graben** [comment]
@gytis : Add missing pooled test esspecially to show the differnence Narayana is handling the XAResource.

**Comment by gytis** [comment]
You're right @graben . My initial test was not correct. I had to refresh my memory of the Narayana internals.

I looked at the Transactional Driver implementation and I don't think that we could make it work with the current LRCO implementation. It is preconfigured to wrap certain database connections with an `IsSameRMOverrideXAResource` and ends up hiding our `LrcoXAResource`.

It's different for the pooled connection, which does not use the Transactional Driver. We can support LRCO there, we just need to update the recovery test to expect DB entry without a recovery. The question is though whether such an LRCO optimisation is useful there?

**Comment by graben** [comment]
Well, both use cases work. But the generic way is not real LRCO and might be a bit more risky because commit may fail while other resources got committed, yet in second phase.

What do you mean with your last question? I think it's useful to have LRCO anyway even if only for pooled resource. The test case should be fine. To avoid potential errors it should be disallowed for generic, maybe?

**Comment by gytis** [comment]
Both cases work, but the generic one does not do LRCO. Even though our resource implements the `LastResourceCommitOptimisation` it is wrapped with IsSameRMOverrideXAResource and when the transaction manager creates its record it does not see the `LastResourceCommitOptimisation`. That's why in the end prepare and commit are performed. I think it would make sense to not allow LRCO with a generic data source wrapper in order to avoid any confusion.

**Comment by graben** [comment]
Okay, I'll prepare an updated patch to avoid generic LRCO with tests and documentation.

**Comment by graben** [comment]
@gytis : Patch supplied ;)

**Review by gytis — APPROVED** [review]
Looks great. Thanks @graben.

**Comment by graben** [comment]
@gytis : Great, do you think we can have a 2.5.1 soon?

**Comment by gytis** [comment]
I don't see why not. I'll try to get it out this week

**Pull request merged**